File: mybox/package/clone.py
from typing import Optional

from pydantic import Field

from ..tracker import Tracker
from ..utils import RunArg, async_cached, repo_version
from .destination import Destination

DEFAULT_REMOTE = "origin"


class Clone(Destination):
    repo: str = Field(..., alias="clone")

    def derive_name(self) -> str:
        return self.repo

    async def directory_exists(self) -> bool:
        return await self.driver.is_dir(await self.destination())

    @async_cached
    async def git_args(self) -> list[RunArg]:
        return ["git", "-C", await self.destination()]

    async def run_git(self, *args: RunArg) -> None:
        await self.driver.run(*await self.git_args(), *args)

    async def run_git_output(self, *args: RunArg) -> str:
        return await self.driver.run_output(*await self.git_args(), *args)

    async def local_version(self) -> Optional[str]:
        if not await self.directory_exists():
            return None
        return await self.run_git_output("rev-parse", "HEAD")

    @property
    def remote(self):
        if self.repo.startswith("https://") or "@" in self.repo:
            return self.repo
        return f"https://github.com/{self.repo}.git"

    async def get_remote_version(self) -> str:
        return await repo_version(self.remote)

    async def branch_name(self, ref: str) -> str:
        return await self.run_git_output("rev-parse", "--abbrev-ref", ref)

    async def install(self, *, tracker: Tracker) -> None:
        destination = await self.destination()

        await self.driver.makedirs(destination.parent)

        if not await self.directory_exists():
            await self.driver.run("git", "clone", self.remote, destination)

        await self.run_git("remote", "set-url", DEFAULT_REMOTE, self.remote)

        default_remote_branch = await self.branch_name(f"{DEFAULT_REMOTE}/HEAD")
        default_branch = default_remote_branch.split("/")[1]
        await self.run_git("fetch", "--no-tags", DEFAULT_REMOTE, default_branch)

        current_branch = await self.branch_name("HEAD")
        if current_branch != default_branch:
            await self.run_git("switch", default_branch)
        await self.run_git("reset", "--hard", default_remote_branch)

        tracker.track(destination, root=self.root)

        await super().install(tracker=tracker)

File: src/Mybox/Package/Clone.hs
module Mybox.Package.Clone (ClonePackage (..), mkClonePackage) where

import Data.Text qualified as Text

import Mybox.Aeson
import Mybox.Driver
import Mybox.Effects
import Mybox.Package.Class
import Mybox.Package.Destination
import Mybox.Package.Post
import Mybox.Package.Queue
import Mybox.Package.System
import Mybox.Prelude
import Mybox.Tracker

data ClonePackage = ClonePackage
  { repo :: Text
  , destination :: Path AnyAnchor
  , branch :: Maybe Text
  , post :: [Text]
  }
  deriving (Eq, Generic, Show)

instance HasField "root" ClonePackage Bool where
  getField _ = False

mkClonePackage :: Text -> Path AnyAnchor -> ClonePackage
mkClonePackage repo destination =
  ClonePackage
    { repo
    , destination
    , branch = Nothing
    , post = []
    }

instance HasField "name" ClonePackage Text where
  getField p = Text.intercalate "#" [p.repo, p.destination.text]

instance PackageName ClonePackage where
  withoutName = genericWithoutName' ["repo", "destination"]

instance FromJSON ClonePackage where
  parseJSON = withObjectTotal "ClonePackage" $ do
    repo <- takeField "clone"
    branch <- takeFieldMaybe "branch"
    destination <- takeField "destination"
    post <- takePost
    pure ClonePackage{..}

instance ToJSON ClonePackage where
  toJSON p =
    object $
      [ "clone" .= p.repo
      , "branch" .= p.branch
      , "destination" .= p.destination
      ]
        <> postToJSON p

prerequisites :: App es => Eff es ()
prerequisites = queueInstall $ mkSystemPackage "git"

cpRemote :: ClonePackage -> Text
cpRemote p
  | Text.isPrefixOf "https://" r = r
  | Text.isInfixOf "@" r = r
  | otherwise = "https://github.com/" <> r <> ".git"
 where
  r = p.repo

cpGitArgs :: Driver :> es => [Text] -> ClonePackage -> Eff es (NonEmpty Text)
cpGitArgs args p = do
  dest <- destinationPath p
  pure $ "git" :| (["-C", dest.text] <> args)

cpRunGit :: Driver :> es => [Text] -> ClonePackage -> Eff es ()
cpRunGit args p = cpGitArgs args p >>= drvRun

cpRevParse ::
  Driver :> es =>
  -- | branch name
  Text ->
  -- | abbrev-ref
  Bool ->
  ClonePackage ->
  Eff es Text
cpRevParse branch abbrevRef p =
  cpGitArgs (join [["rev-parse"], ["--abbrev-ref" | abbrevRef], [branch]]) p
    >>= drvRunOutput

cpLocalVersion :: App es => ClonePackage -> Eff es (Maybe Text)
cpLocalVersion p = do
  exists <- destinationExists p
  if exists
    then do
      prerequisites
      Just <$> cpRevParse "HEAD" False p
    else pure Nothing

cpRemoteVersion :: App es => ClonePackage -> Eff es Text
cpRemoteVersion p = do
  prerequisites
  drvRepoBranchVersion (cpRemote p) p.branch

cpDefaultRemote :: Text
cpDefaultRemote = "origin"

cpRemoteBranch :: Text -> Text
cpRemoteBranch b = cpDefaultRemote <> "/" <> b

cpInstall :: App es => ClonePackage -> Eff es ()
cpInstall p = do
  prerequisites
  destination <- destinationPath p
  exists <- drvIsDir $ destination </> ".git"
  if exists
    then cpRunGit ["remote", "set-url", cpDefaultRemote, cpRemote p] p
    else do
      drvMkdir destination.dirname
      drvRm destination
      drvRun $ "git" :| ["clone", cpRemote p, destination.text]
  remoteBranch <-
    case p.branch of
      Just branch -> pure $ cpRemoteBranch branch
      Nothing -> cpRevParse (cpRemoteBranch "HEAD") True p
  let branch = Text.takeWhileEnd (/= '/') remoteBranch
  cpRunGit ["fetch", "--no-tags", cpDefaultRemote, branch] p
  currentBranch <- cpRevParse "HEAD" True p
  when (currentBranch /= branch) $ cpRunGit ["switch", branch] p
  trkAdd p destination
  cpRunGit ["reset", "--hard", remoteBranch] p

instance Package ClonePackage where
  localVersion = cpLocalVersion
  remoteVersion = cpRemoteVersion
  install = installWithPost cpInstall
